File: core/src/nl/t64/game/rpg/screens/world/mapobjects/GameMapEventDiscover.java
package nl.t64.game.rpg.screens.world.mapobjects;

import com.badlogic.gdx.maps.objects.RectangleMapObject;
import com.badlogic.gdx.math.Rectangle;
import nl.t64.game.rpg.Utils;
import nl.t64.game.rpg.components.event.Event;
import nl.t64.game.rpg.screens.world.entity.Direction;
import nl.t64.game.rpg.subjects.CollisionObserver;


public class GameMapEventDiscover extends GameMapObject implements CollisionObserver {

    private final String eventId;

    public GameMapEventDiscover(RectangleMapObject rectObject) {
        super.rectangle = rectObject.getRectangle();
        this.eventId = rectObject.getName();

        Utils.getBrokerManager().collisionObservers.addObserver(this);
    }

    @Override
    public void onNotifyCollision(Rectangle playerBoundingBox, Direction playerDirection) {
        if (playerBoundingBox.overlaps(rectangle)) {
            possibleStartEvent();
        }
    }

    private void possibleStartEvent() {
        Event event = Utils.getGameData().getEvents().getEventById(eventId);
        event.possibleStart();
    }

}

File: core/src/nl/t64/game/rpg/screens/world/mapobjects/GameMapEventDiscover.kt
package nl.t64.game.rpg.screens.world.mapobjects

import com.badlogic.gdx.maps.objects.RectangleMapObject
import com.badlogic.gdx.math.Rectangle
import nl.t64.game.rpg.Utils.brokerManager
import nl.t64.game.rpg.Utils.gameData
import nl.t64.game.rpg.screens.world.entity.Direction
import nl.t64.game.rpg.subjects.CollisionObserver


class GameMapEventDiscover(rectObject: RectangleMapObject) : GameMapObject(rectObject.rectangle), CollisionObserver {

    private val eventId: String = rectObject.name

    init {
        brokerManager.collisionObservers.addObserver(this)
    }

    override fun onNotifyCollision(playerBoundingBox: Rectangle, playerDirection: Direction) {
        if (playerBoundingBox.overlaps(rectangle)) {
            possibleStartEvent()
        }
    }

    private fun possibleStartEvent() {
        val event = gameData.events.getEventById(eventId)
        event.possibleStart()
    }

}
